bpo-34155: Dont parse domains containing @

jpic opened the pull request:
Before:
    
        >>> email.message_from_string('From: a@malicious.org@important.com', policy=email.policy.default)['from'].addresses
        (Address(display_name='', username='a', domain='malicious.org'),)
    
        >>> parseaddr('a@malicious.org@important.com')
        ('', 'a@malicious.org')
    
    After:
    
        >>> email.message_from_string('From: a@malicious.org@important.com', policy=email.policy.default)['from'].addresses
        (Address(display_name='', username='', domain=''),)
    
        >>> parseaddr('a@malicious.org@important.com')
        ('', 'a@')



<!-- issue-number: [bpo-34155](https://bugs.python.org/issue34155) -->
https://bugs.python.org/issue34155
<!-- /issue-number -->


Automerge-Triggered-By: @warsaw

maxking commented on Lib/email/_header_value_parser.py:
```diff
@@ -1559,6 +1559,8 @@ def get_domain(value):
     except errors.HeaderParseError:
         token, value = get_atom(value)
+    if value and value[0] == '@':
+        raise errors.HeaderParseError('Multiple domains')
```
I'd rename the error as `Invalid Domain`.

maxking commented on Lib/test/test_email/test_email.py:
```diff
@@ -3032,6 +3032,20 @@ def test_parseaddr_empty(self):
+    def test_parseaddr_multiple_domains(self):
+        self.assertEqual(
+            utils.parseaddr('a@b@c'),
+            ('', 'a@')
```
According to the docs:
> email.utils.parseaddr(address)
>    Parse address – which should be the value of some address-containing field such as To or Cc – into its constituent realname and email address parts. Returns a tuple of that information, unless the parse fails, in which case a 2-tuple of ('', '') is returned.

Can we expect the same in this case?

I made this change to make that happen:
```
@@ -379,7 +379,12 @@ class AddrlistClass:
         aslist.append('@')
         self.pos += 1
         self.gotonext()
-        return EMPTYSTRING.join(aslist) + self.getdomain()
+        domain = self.getdomain()
+        if not domain:
+            # Invalid domain, return an empty address instead of returning a
+            # local part to denote failed parsing.
+            return EMPTYSTRING
+        return EMPTYSTRING.join(aslist) + domain
```

jpic replied:
Sure, please confirm that I got it right:

You mean that `utils.parseaddr('a@b@c')` should return `('', '')`, or just `''` ?

Just one question about consistency, would that change also implicate a change of the following : 

```
>>> parseaddr('foo@a.') 
('', 'foo@a.')
```

Should it also return same as with erroneous caracter @ in domain ? (an empty string or tuple of empty strings)

If so, should that also be in the security patch or do you want a new issue and a new patch for breaking backward compatibility ?

To be completely honest with you, I was convinced this patch would pass because I didn't see any other implementation that I could propose on my own in the context of closing the security issue. I see that there is more work to keep on doing in the email library and am looking forward to do more, in this patch request or another, I'm all yours for homework assignments @maxking.

Thank you for your review !

maxking replied:
Yes, I meant return value should be `('', '')`.

Coming to the behavior of `parseaddr`, here is my thinking on this this issue. Right now, we return either "whatever we can parse" (which is why we return the invalid domains as-is without opinions) and `('', '')` when there is an exception raised during parsing.

So, following this line of reasoning, with domains like `a@b@c.com` should actually return `('', ,'a@b@c.com' )`. I think this is the right thing to do. *If* you think this behavior can cause security issue for the downstreams who might have similar bug and would take in `a@b@c.com` as `a@b` and think we should consider this, then the next right thing to do would be to return an empty string denoting unparseable address.

Does that line of reasoning made sense like it did in my head?

jpic replied:
Yes of course it is expected to break BC to fix security issue. In this case do you want this patch to also change parseaddr('a@b.') to return a tuple of two empty strings in this patch for consistency with parseaddr('a@b@c') which is now going to return an empty strings tuple ? Sorry you make sense but there i am not sure if this consistency point is important or not for the time being

jpic replied:
I think we can deal with this later, so I added your patch, thanks @maxking ! But I didn't know how to add credit

maxking replied:
You can add credits in commit message, I think.

You may need to update your tests to reflect the patch too, they seem to be failing right now with new updates.

maxking commented on Misc/NEWS.d/next/Security/2019-05-04-13-33-37.bpo-34155.MJll68.rst:
```diff
@@ -0,0 +1 @@
+Don't parse email domain containing an at, ie. a@malicious.org@important.com
```
I would suggest a slightly different wording for this:

```
Fix parsing of invalid email addresses with more than one ``@`` (e.g. a@b@c.com.) to not return the part before 2nd ``@`` as valid email address. Patch by jpic.
```

maxking reviewed (commented):
Thanks for the patch!

I have some different ideas on how to do error recovery in this case, please see [my comment](https://bugs.python.org/msg344030) on bpo-34155.

warsaw commented on Lib/email/_header_value_parser.py:
```diff
@@ -1584,6 +1584,8 @@ def get_domain(value):
     except errors.HeaderParseError:
         token, value = get_atom(value)
+    if value and value[0] == '@':
+        raise errors.HeaderParseError('Invalid Domain')
```
I wonder if this should add a defect instead of raising a `HeaderParseError`?  I can't decide.  I suppose the exception is okay as long as it doesn't get exposed in the higher level email parsing APIs (e.g. `message_from_string()`, since those are not allowed to raise exceptions.

jpic replied:
The [HeaderParseError is expected by the higher level API which adds a defect](https://github.com/python/cpython/blob/02257012f6d3821d816cb6a7e8461a88a05b9a08/Lib/email/_header_value_parser.py#L1956) so you don't have to take any decision. As a user you can still know about that parse problem using the defects attribute:
```
>>> email.message_from_string('From: a@malicious.org@example.com',
policy=email.policy.default)['from'].defects[0]
InvalidHeaderDefect('invalid address in address-list')
```
https://bugs.python.org/msg341381

warsaw commented on Lib/email/_parseaddr.py:
```diff
@@ -394,6 +400,9 @@ def getdomain(self):
                 self.pos += 1
                 sdlist.append('.')
+            elif self.field[self.pos] == '@':
+                #34155 Don't parse domain for a@malicious.org@important.com
```
Please label this `# bpo-34155` since it has to be unique in case we eventually do change bug trackers.

jpic replied:
Is this version ok ?

https://github.com/python/cpython/pull/13079/files#diff-18392c73bfdad73b43ab16d2adfd5b1eR403

jpic committed:
bpo-34155: Fail parse of multiple email domains

Before:

    >>> email.message_from_string('From: a@malicious.org@important.com', policy=email.policy.default)['from'].addresses
    (Address(display_name='', username='a', domain='malicious.org'),)

    >>> parseaddr('a@malicious.org@important.com')
    ('', 'a@malicious.org')

After:

    >>> email.message_from_string('From: a@malicious.org@important.com', policy=email.policy.default)['from'].addresses
    (Address(display_name='', username='', domain=''),)

    >>> parseaddr('a@malicious.org@important.com')
    ('', '')

Thanks @maxking for kind patch review and patch contribution.

maxking commented on Lib/email/_parseaddr.py:
```diff
@@ -394,6 +399,9 @@ def getdomain(self):
                 self.pos += 1
                 sdlist.append('.')
+            elif self.field[self.pos] == '@':
+                #34155 Don't parse domain for a@malicious.org@important.com
```
Can you please make this
```
# bpo-34155: Don't parse domains with two `@` like `a@malicious.org@important.com`.
```

We prefer to use `bpo-` to link to issues in the code, which stands for bugs.python.org :)

jpic replied:
The comment was made to fit without exceeding 80 caracters :joy:

Would this work for you ?

```
                # bpo-34155: Don't parse domains with two `@` like
                # `a@malicious.org@important.com`.
```

Or

```
# bpo-34155: Don't parse two @ like a@malicious@ok.com
```

maxking replied:
I think two lines are fine, we don't really need to fit it into a single line.

jpic replied:
Thanks for your reply it's pushed !

maxking reviewed (commented):
The patch looks good to me, just a very teeny tiny nitpick that I have added inline comment for.

@warsaw I think after the fix for the comment, this patch should be good to go! 
Needs backport to 3.6, 3.7, 3.8 since this has security implications.
